### DISCUSSION TITLE
Limit max request size

### DISCUSSION THREAD

**Discussion by aviramha** [encode/starlette#1516] [Ideas]
As discussed in the Gitter, my opinion is that starlette should provide a default limit for request size.
The main reason is that without it, any Starlette application is vulnerable to very easy DoS.
For example, newbie me can write a program as follows:
```
from starlette.requests import Request
from starlette.responses import Response


async def app(scope, receive, send):
    assert scope['type'] == 'http'
    request = Request(scope, receive)
    body = b''
    json = await request.json()
    response = Response(body, media_type='text/plain')
    await response(scope, receive, send)
```
As a malicious user, I could send a 30GB sized JSON and cause the memory to go OOM.
Other frameworks support this also - Django, Quart.
My proposal is to add a default limit which can be overrided in the app configuration.

**Comment by florimondmanca** [comment]
As far as I intuitively agree with the relevance of supporting this in Starlette core, my two cents here...

> My proposal is to add a default limit which can be overrided in the app configuration.

Would this be a kwarg on the `Starlette` class? I don’t think there’s precedent for this - everything else is provided as middleware (even security features such as CORS or trusted hosts). Also not sure this would be the most reusable API (whereas other frameworks would benefit from a public ASGI middleware).

So personally not sure we want a new config option, or just a middleware that’s documented as « you should _really_ turn this on for production » (both in the middleware docs and any « deployment » docs).

**Comment by aviramha** [comment]
> As far as I intuitively agree with the relevance of supporting this in Starlette core, my two cents here...
> 
> > My proposal is to add a default limit which can be overrided in the app configuration.
> 
> Would this be a kwarg on the `Starlette` class? I don’t think there’s precedent for this - everything else is provided as middleware (even security features such as CORS or trusted hosts). Also not sure this would be the most reusable API (whereas other frameworks would benefit from a public ASGI middleware).
> 
> So personally not sure we want a new config option, or just a middleware that’s documented as « you should _really_ turn this on for production » (both in the middleware docs and any « deployment » docs).

What about a middleware that's used by default? For example there's already:
```
middleware - A list of middleware to run for every request. A starlette application will always automatically include two middleware classes. ServerErrorMiddleware is added the very outermost middleware, to handle any uncaught errors occuring anywhere in the entire stack. ExceptionMiddleware is added as the very innermost middleware, to deal with handled exception cases occuring in the routing or endpoints.
```
Add to this list the request size middleware. Another option is that the middleware parameter has default value that has the middleware.

**Comment by florimondmanca** [comment]
Including the middlware by default leads to the question of how to leave it out in case users want to specify it explicitly… `ServerErrorMiddleware`/`ExceptionMiddleware` work well that way because they don't have any associated config (apart from `exception_handlers`).

I don't think it'd be terribly awful for Starlette to not have everything turned-on-and-secure-by-default, i.e. if there are small steps required for the developer to make their app production-ready. (The philosophy is that of an ASGI toolkit and _lightweight_ framework, so that's how I interpret it.)

There's been related discussion on FastAPI too, eg https://github.com/tiangolo/fastapi/issues/362. Conversation was resolved as "you can implement this as middleware", and that feature was not provided in core. (Always a good idea to go see how things are going there, as a lot of FastAPI users want to see stuff in FastAPI whereas it would really benefit from being in Starlette or separate ASGI components…)

So I'm still inclined to think that "here's an upload limit ASGI middleware, please turn it on for prod" (i.e. no default, but easy to add) would be an okay approach for Starlette and the wider ASGI ecosystem.

**Comment by aviramha** [comment]
What about optional argument to the json, body and form functions with default values?
i.e
```
    async def json(self, max_size: int = 64000) -> typing.Any:
        if not hasattr(self, "_json"):
            body = await self.body(max_size)
            self._json = json.loads(body)
        return self._json
```

**Comment by rafalp** [comment]
Wouldn't it be more reliable for this limit to live on edge server, such as NGINX?

**Comment by aviramha** [comment]
> Wouldn't it be more reliable for this limit to live on edge server, such as NGINX?

It is possible to limit it using NGINX, but you don't want to couple Starlette's resilience with an edge server.

**Comment by erewok** [comment]
I always have these limits in Nginx. It would not occur to me to have these limits in the framework but others may need this if they're exposing services without a reverse proxy in front? I'd just want a way to turn it off or not have it enabled by default.

**Comment by aviramha** [comment]
> I always have these limits in Nginx. It would not occur to me to have these limits in the framework but others may need this if they're exposing services without a reverse proxy in front? I'd just want a way to turn it off or not have it enabled by default.

Because if Starlette doesn't need a reverse proxy, why add it? if Starlette relies on a reverse proxy to be protected it needs to be well documented and IMO it's a bad decision, as each layer need to protect itself.
Layering features is smart and OK, but layering protection is not really smart.. (Especially when the cost of adding this protection is so trivial and minimal)

**Comment by erewok** [comment]
> Because if Starlette doesn't need a reverse proxy, why add it?

I deploy everything behind a reverse proxy because my services require load balancing and redundancy. 

In fairness, however, I was trying to remember what Django does for this, and so I went looking for it. I honestly can't ever remember [setting this `DATA_UPLOAD_MAX_MEMORY_SIZE` parameter](https://docs.djangoproject.com/en/3.0/ref/settings/#data-upload-max-memory-size) in the past but I must have because I've had apps that accepted large file uploads, much larger than the defaulf of 2.5MB described in their docs.

There's a similar discussion on [this issue in Gunicorn](https://github.com/benoitc/gunicorn/issues/1659) where in addition to talking about DDOS attacks, they also talk about what Werkzeug (the library underpinning Flask) [does](https://werkzeug.palletsprojects.com/en/1.0.x/request_data/#limiting-request-data). 

Further, it appears that Tornado also offers a `max_buffer_size` argument to their `HTTPServer` class, which _closes_ the connection when a request exceeds that size (by default, it's set to 100MB).

In summary, it seems like lots of Python web frameworks (although I tend of think of Tornado as more of a server) implement something like this, but there's some concern about trending too far into full-blown server territory because there are could be many strategies to mitigate DDOS attacks that belong elsewhere than the web framework. Thus, it's probably not an uncontroversial thing to add if all the established libraries already offer it.

**Comment by gnat** [comment]
IMO there really should be a configurable limit and reasonable default if its not too expensive to do a simple check. This is an attack vector everyone has by default.

Right now I feel stuff like this keeps Starlette as more of a "pro only" framework rather than something usable by any level of developer.

**Comment by Chaostheorie** [comment]
Has there been any progress so far on this? Or should I configure my webserver or uvicorn to do this?

**Comment by gnat** [comment]
For those following this, even behind nginx, I've realised there's a number of ways to run denial of service on Starlette which can be mitigated with a simple timeout middleware. I recommend a dual solution of setting up nginx `client_max_body_size`, and using a Starlette middleware.

1. For nginx, you need to set  `client_max_body_size` to limit total body size (and in turn, file uploads). For example, to limit to 100 MB, your core http config block should look like: 
```
http {
    client_max_body_size 100m;
    # ... other stuff ...
}
```

2. As for Starlette itself, a timeout middleware can serve as a "catch all" solution for large file uploads and long running queries. Sample middleware here: https://github.com/tiangolo/fastapi/issues/1752#issuecomment-682579845

```python
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import HTMLResponse
import asyncio

class TimeoutMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        try:
            response = await asyncio.wait_for(call_next(request), timeout=30)
        except asyncio.TimeoutError:
            return HTMLResponse("Request reached timeout.", status_code=504)
        return response
```

Timeout middleware seems to free the memory used by the attempted file upload as well.

Any additional thoughts, suggestions appreciated! We all want to achieve the goal of making our Starlette Apps more robust.

**Comment by gnat** [comment]
Note: By default PHP sets a request time limit of 30 seconds since version 4.0, which undoubtedly handles a lot of DOS sidecases.

https://www.php.net/manual/en/function.set-time-limit.php

**Comment by gnat** [comment]
2023 note: Sanic Server has these safeties out of the box, if you're looking for something that just works.

Still no sane default in the Uvicorn / Starlette stack, even with nginx in front- You must use the middleware or risk a DOS.

**Comment by gnat** [comment]
2025 note: [Tremolo](https://github.com/nggit/tremolo) another minimalist ASGI framework + server has:
* Request timeouts.
* Bandwidth rate limits.
* Max body/header sizes.
* Max memory limits.

**Comment by dacevedo12** [comment]
Any plans of adding this to starlette? Looks like a sane default

**Reply by nggit** [reply]
It is not a necessity for Starlette to implement that.
But on the other hand, its server, Uvicorn has yet to have any further constraints.

As [someone already reviewed](https://github.com/encode/starlette/discussions/1516#discussioncomment-12111087), you can use Tremolo (ASGI server mode) for now to run Starlette.
I'm too tired to reveal it to the public as it's not the fastest. But there is a lot of concern for abusers. Even bad applications like the following are not allowed:
```python
# viva:app
import asyncio


async def app(scope, receive, send):
    assert scope['type'] == 'http'

    await send({
        'type': 'http.response.start',
        'status': 200,
        'headers': [
            (b'content-type', b'text/plain'),
            (b'connection', b'close'),
        ]
    })
    await send({
        'type': 'http.response.body',
        'body': b'Hello, World! (%d)' % len(asyncio.all_tasks()),
    })
    # <-- connection_lost

    # viva forever
    await asyncio.get_running_loop().create_future()
```

Every time you reload the browser, you will see the number go up without going down. This is not the case with Tremolo, nor Daphne.

**Reply by gnat** [reply]
> I'm too tired to reveal it to the public as it's not the fastest.

Slightly off topic; I'm shocked Tremolo does not get more natural attention in this community, I feel it's one of the best in this category on Github. Tiny, feature rich, true zero dependency ASGI, all-in-one. It's a true gem!!

Performance is nice but super overrated because web servers should horizontally scale. Simplicity, maintainability, not crashing, and not vulnerable to DDOS all become a lot more valuable.

Remember Github is just a bunch of horizontal stateless Rails servers (using 1 master MySQL up until only a few years ago, now Vitess).

**Reply by nggit** [reply]
> Performance is nice but super overrated because web servers should horizontally scale.

Because it can take several years to reach this conclusion.

But since CPython is getting faster, I would guess that it now has similar performance vs expressjs, with Python's productivity benefits.

I allow off topic, but this is not in my domain. Let's move on if you have a lot of thoughts.

**Comment by Raphasha27** [comment]
Starlette doesn't have a built-in max request size setting, but you can enforce it with a simple middleware:

```python
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import PlainTextResponse

class MaxBodySizeMiddleware(BaseHTTPMiddleware):
    def __init__(self, app, max_size: int = 10 * 1024 * 1024):  # default 10MB
        super().__init__(app)
        self.max_size = max_size

    async def dispatch(self, request, call_next):
        content_length = request.headers.get("content-length")
        if content_length and int(content_length) > self.max_size:
            return PlainTextResponse(
                "Request too large",
                status_code=413,
            )
        # For streaming bodies, cap the size
        if request.method in ("POST", "PUT", "PATCH"):
            body = await request.body()
            if len(body) > self.max_size:
                return PlainTextResponse(
                    "Request too large",
                    status_code=413,
                )
        return await call_next(request)

# Usage
app.add_middleware(MaxBodySizeMiddleware, max_size=5 * 1024 * 1024)  # 5MB
```

If you're using FastAPI, you can also set `max_request_size` via uvicorn:

```bash
uvicorn app.main:app --limit-max-request-body 5242880  # 5MB in bytes
```

Note that `request.body()` reads the entire body into memory, so for very large file uploads, consider using `Stream` or checking `content-length` header first (which is what the middleware above does).

**Comment by ArjunTomar1402** [comment]
I'd like to revisit this discussion.

I noticed most of the previous conversation focused on middleware versus application configuration.

While experimenting with the current Request implementation, I realized every body-reading API (`body()`, `json()`, `form()`, and manual streaming) ultimately goes through `Request.stream()`.

Instead of wrapping `receive()` in middleware, would maintainers consider exposing a configurable limit through `Request.stream()` (for example via a namespaced scope key set by middleware)?

This would keep the configuration in middleware while centralizing enforcement at the point where Starlette actually consumes request bytes.

I'd be happy to prototype this if the overall direction seems reasonable.
